buildSrc/src/main/java/org/mozilla/gradle/tasks/ValidateAndroidAppReleaseConfiguration.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.gradle.tasks
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+import org.mozilla.gradle.GitAggregates
+import org.mozilla.gradle.ext.androidDSLOrThrow
+
+/**
+ * Validates that an Android app is correctly configured for release. When created, this
+ * task will automatically create dependencies upon the release tasks.
+ *
+ * This task currently asserts:
+ * - There are no uncommitted changes in git
+ * - The buildVersion in gradle matches the checked out git tag
+ *
+ * This task can be skipped by supplying the `-Pno-validate` argument.
+ */
+open class ValidateAndroidAppReleaseConfiguration : DefaultTask() {
+    init {
+        group = "Verification"
+        description = "Validates an Android app is correctly configured for release"
+
+        this.onlyIf { !project.hasProperty("no-validate") }
+
+        project.gradle.projectsEvaluated {
+            getAssembleReleaseTasks().forEach {
+                it.dependsOn(this@ValidateAndroidAppReleaseConfiguration)
```

Comment:
Hooking into build order seems unrelated to this class' main responsibility (validating git and version line up).  Suggest breaking these out of this class and putting them into extension functions.  They seem fairly likely to be reused in other contexts.
```
project.assembleReleaseDependsOn(this)
project.compileReleaseMustRunAfter(this)
```

Reply:
I don't think hooking into build order is out of scope – a task is its execution and non-parameterized configuration –  but I think it's out of scope to iterate through the tasks to determine what an "assembleRelease" task is. It also pains me how overly specific a method like, `project.assembleReleaseDependsOn(this)` is. As such, I took the middle ground and changed the code to this:
```kotlin
        project.gradle.projectsEvaluated { project.tasks.let { tasks ->
            // The assemble release tasks run late during execution but the compile tasks run
            // early and assemble depends on them so we make our new task run before these compile tasks.
            tasks.getAssembleReleaseTasks().forEach { it.dependsOn(this@ValidateAndroidAppReleaseConfiguration) }
            tasks.getCompileReleaseTasks().forEach { it.mustRunAfter(this@ValidateAndroidAppReleaseConfiguration) }
        } }

// TaskContainer.kt:
fun TaskContainer.getAssembleReleaseTasks(): List<Task> = filter { task -> task.name.let {
    it.startsWith("assemble") && it.endsWith("Release")
} }

fun TaskContainer.getCompileReleaseTasks(): List<Task> = filter { task -> task.name.let {
    it.startsWith("compile") && it.endsWith("ReleaseSources")
} }
```

Reply:
Thanks for always thinking about "whose responsibility it is to do this"!